### ISSUE TITLE
Allowing send + sync for Context

### ISSUE THREAD

**Issue by 0xpr03** [jnqnfe/pulse-binding-rust#14]
I'm not sure whether it's safe to make Context send, but I'm currently failing to share an `Arc<Mutex<Context>>` due to 
```log
`*const extern "C" fn(*mut pulse_sys::pa_context, *const i8, *mut pulse_sys::pa_proplist, *mut std::ffi::c_void)` cannot be sent between threads safely
```

I'm not sure if it helps, this is the full log output when trying to do:
`(move || ) the context Arc<Mutex<_>>`

<details>

```log
help: within `pulse::context::Context`, the trait `std::marker::Send` is not implemented for `*const extern "C" fn(*mut pulse_sys::pa_context, *const i8, *mut pulse_sys::pa_proplist, *mut std::ffi::c_void)`
   = note: required because it appears within the type `std::marker::PhantomData<*const extern "C" fn(*mut pulse_sys::pa_context, *const i8, *mut pulse_sys::pa_proplist, *mut std::ffi::c_void)>`
   = note: required because it appears within the type `pulse::callbacks::MultiUseCallback<(dyn std::ops::FnMut(std::string::String, pulse::proplist::Proplist) + 'static), extern "C" fn(*mut pulse_sys::pa_context, *const i8, *mut pulse_sys::pa_proplist, *mut std::ffi::c_void)>`
   = note: required because it appears within the type `pulse::context::CallbackPointers`
   = note: required because it appears within the type `pulse::context::Context`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Mutex<pulse::context::Context>`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<std::sync::Mutex<pulse::context::Context>>`
```

</details>

I'm trying to share a Context & Mainloop with a hyper thread, both wrapped inside an Arc<Mutex<_>>, so my only option for now would be to handle these in a new thread and use an mpsc channel to communicate between them.

**Comment by 0xpr03** [comment]
I've made my struct send + sync and have no problem, so it should be okay to add this to mainloop & context.

**Comment by jnqnfe** [comment]
Ok, I have not done a properly thorough investigation, but I have evaluated it and spend a fair bit of time coming through various bits and pieces. With the threaded mainloop you need to hold the lock while using objects, thus so long as you do that, I believe everything should be fine, as discussed in the relevant commit I have made and its changelog entry. I have released v2.5 with the necessary change :)

Also, I found a use-after-free bug whilst looking into this, so thanks for prompting me to do so :)

**Comment by 0xpr03** [comment]
Hi, many thanks for investigating.
I'm running my hotfix stuff inside a tokio engine between multiple hyper instances and have not faced any issue for far.
Regarding your problem of users forgetting to acquire a lock:
AFAIK you could make the methods require `&mut self` instead of `&self`, this way borrowck would enforce using a mutex.
